File: src/sjekklista-hr/Sjekklista.Hr.ApiService.Tests.Integration/Features/VacationPlanning/VacationPlanningClient.cs
using Sjekklista.Hr.ApiService.Features.VacationPlanning.Contracts;

namespace Sjekklista.Hr.ApiService.Tests.Integration.Features.VacationPlanning
{
    public class VacationPlanningClient
    {
        private readonly HttpClient _httpClient;

        public VacationPlanningClient(Guid tenantId, WebApplicationFactoryFixture fixture)
        {
            _httpClient = fixture.Factory.CreateClient();
            _httpClient.DefaultRequestHeaders.Add("X-Tenant", tenantId.ToString());
        }

        internal async Task<HttpResponseMessage> GetEmployeeVacationPlan(Guid employeeId, int year)
        {
            var response = await _httpClient.GetAsync($"/api/vacation-planning/employee/{year}/{employeeId}");
            return response;
        }

        internal async Task<HttpResponseMessage> SaveEmployeeVacationPlan(
            SaveEmployeeVacationPlanRequest request)
        {
            return await _httpClient.PostAsJsonAsync("/api/vacation-planning/employee", request);
        }

        internal async Task<HttpResponseMessage> RequestVacation(
            RequestVacationRequest request)
        {
            return await _httpClient.PostAsJsonAsync("/api/vacation-planning/employee/request-vacation", request);
        }

        internal async Task<HttpResponseMessage> SetupVacation(
            SetupVacationRequest request)
        {
            return await _httpClient.PostAsJsonAsync("/api/vacation-planning/employee/setup-vacation", request);
        }

        internal async Task<HttpResponseMessage> GetAllVacationPlans(int year)
        {
            return await _httpClient.GetAsync($"/api/vacation-planning/employee/year/{year}");
        }

        internal async Task<HttpResponseMessage> InitializeVacationPlans(
            InitializeVacationPlansRequest request)
        {
            return await _httpClient.PostAsJsonAsync("/api/vacation-planning/initialize", request);
        }
    }
}

File: src/sjekklista-hr/sjekklistahr/src/data/clients/vacationPlanningClient.ts
import { apiClient } from '../apiClient';
import type {
  EmployeeVacationPlanDto,
  GetEmployeeVacationPlanResponse,
  RequestVacationRequest,
  RequestVacationResponse,
  SaveEmployeeVacationPlanRequest,
  SaveEmployeeVacationPlanResponse,
  SetupVacationRequest,
  SetupVacationResponse,
} from '../models/VacationPlanningModels';

class VacationPlanningClient {
  private readonly basePath = '/vacation-planning';

  async getEmployeeVacationPlan(
    employeeId: string,
    year: number
  ): Promise<EmployeeVacationPlanDto> {
    const response = await apiClient
      .getHttpClient()
      .get<GetEmployeeVacationPlanResponse>(
        `${this.basePath}/employee/${year}/${employeeId}`
      );
    return response.data.employeeVacation;
  }

  async saveEmployeeVacationPlan(
    request: SaveEmployeeVacationPlanRequest
  ): Promise<EmployeeVacationPlanDto> {
    const response = await apiClient
      .getHttpClient()
      .post<SaveEmployeeVacationPlanResponse>(
        `${this.basePath}/employee`,
        request
      );
    return response.data.employeeVacation;
  }

  async requestVacation(
    request: RequestVacationRequest
  ): Promise<EmployeeVacationPlanDto> {
    const response = await apiClient
      .getHttpClient()
      .post<RequestVacationResponse>(
        `${this.basePath}/employee/request-vacation`,
        request
      );
    return response.data.employeeVacation;
  }

  async setupVacation(
    request: SetupVacationRequest
  ): Promise<EmployeeVacationPlanDto> {
    const response = await apiClient
      .getHttpClient()
      .post<SetupVacationResponse>(
        `${this.basePath}/employee/setup-vacation`,
        request
      );
    return response.data.employeeVacation;
  }

  async getAllEmployeeVacationPlans(year: number): Promise<EmployeeVacationPlanDto[]> {
    try {
      const response = await apiClient
        .getHttpClient()
        .get<EmployeeVacationPlanDto[]>(
          `${this.basePath}/employee/year/${year}`
        );
      return response.data;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error) {
        const axiosError = error as { response?: { status?: number } };
        if (axiosError.response?.status === 404) {
          return [];
        }
      }
      throw error;
    }
  }
}

export const vacationPlanningClient = new VacationPlanningClient();
